File: internal/ignore/ignore.go
// Package ignore implements the two-tier suppression mechanism:
//
//  1. Inline comment on the same line or the line immediately above a finding:
//     # tf-firewall-ignore: unknown_attribute,tutorial_pattern
//
//  2. Global list in config.yml (ignore_rules) that suppresses a category
//     across all files.
package ignore

import (
	"bufio"
	"bytes"
	"strings"

	"github.com/foadtalsi/tf-predeploy-firewall/internal/report"
)

const directivePrefix = "tf-firewall-ignore:"

// ParseComments scans raw .tf source and returns a map of
// 1-indexed line number -> set of categories suppressed on that line.
// A directive on line N suppresses findings on line N and line N+1,
// so the caller can place the comment either on the attribute line itself
// or on the line above it.
func ParseComments(src []byte) map[int]map[report.Category]bool {
	out := make(map[int]map[report.Category]bool)
	scanner := bufio.NewScanner(bytes.NewReader(src))
	lineNum := 0
	for scanner.Scan() {
		lineNum++
		line := scanner.Text()
		idx := strings.Index(line, "#")
		if idx < 0 {
			continue
		}
		comment := strings.TrimSpace(line[idx+1:])
		if !strings.HasPrefix(comment, directivePrefix) {
			continue
		}
		cats := parseCategoryList(strings.TrimPrefix(comment, directivePrefix))
		// Suppress on this line and the next (directive above the attribute).
		for _, n := range []int{lineNum, lineNum + 1} {
			if out[n] == nil {
				out[n] = make(map[report.Category]bool)
			}
			for _, c := range cats {
				out[n][c] = true
			}
		}
	}
	return out
}

func parseCategoryList(s string) []report.Category {
	var cats []report.Category
	for _, part := range strings.Split(s, ",") {
		part = strings.TrimSpace(part)
		if part != "" {
			cats = append(cats, report.Category(part))
		}
	}
	return cats
}

// Apply removes findings that are suppressed either by an inline directive
// in their source file or by the global ignore list.
func Apply(
	findings []report.Finding,
	inlineByFile map[string]map[int]map[report.Category]bool,
	globalIgnore []report.Category,
) []report.Finding {
	globalSet := make(map[report.Category]bool, len(globalIgnore))
	for _, c := range globalIgnore {
		globalSet[c] = true
	}

	var out []report.Finding
	for _, f := range findings {
		if globalSet[f.Category] {
			continue
		}
		if fileMap, ok := inlineByFile[f.File]; ok {
			if lineMap, ok := fileMap[f.Line]; ok {
				// "all" suppresses every category on this line
				if lineMap[report.Category("all")] || lineMap[f.Category] {
					continue
				}
			}
		}
		out = append(out, f)
	}
	return out
}

File: src/tfpdf/ignore.py
"""Inline, category, and path-based finding suppression."""

from __future__ import annotations

import posixpath
import re
from collections.abc import Iterable, Sequence
from dataclasses import dataclass, field
from functools import lru_cache

from .report.finding import Category, Finding

DIRECTIVE_PREFIX = "tf-firewall-ignore:"

# Pseudo-category that suppresses all findings on a line.
_ALL = "all"


def parse_comments(source: bytes) -> dict[int, set[str]]:
    """Map one-based source lines to categories suppressed by inline directives."""
    out: dict[int, set[str]] = {}
    text = source.decode("utf-8", errors="replace")
    for line_num, line in enumerate(text.split("\n"), start=1):
        index = line.find("#")
        if index < 0:
            continue
        comment = line[index + 1 :].strip()
        if not comment.startswith(DIRECTIVE_PREFIX):
            continue
        cats = _parse_category_list(comment[len(DIRECTIVE_PREFIX) :])
        # Suppress this line and the next so directives can sit above an attribute.
        for n in (line_num, line_num + 1):
            out.setdefault(n, set()).update(cats)
    return out


def _parse_category_list(s: str) -> list[str]:
    return [part.strip() for part in s.split(",") if part.strip()]


def apply(
    findings: Iterable[Finding],
    inline_by_file: dict[str, dict[int, set[str]]],
    global_ignore: Sequence[Category | str],
) -> list[Finding]:
    """Remove findings covered by inline directives or global category exclusions."""
    global_set = {str(c) for c in global_ignore}

    out: list[Finding] = []
    for finding in findings:
        if str(finding.category) in global_set:
            continue
        line_map = inline_by_file.get(finding.file, {}).get(finding.line)
        if line_map is not None and (_ALL in line_map or str(finding.category) in line_map):
            continue
        out.append(finding)
    return out


@dataclass(slots=True)
class PathRule:
    """A path exclusion supporting ** across zero or more segments, plus * and ? within a segment."""

    pattern: str
    # Accept built-in categories and custom:<id> strings from configuration.
    categories: list[Category | str] = field(default_factory=list)

    def suppresses(self, category: Category | str) -> bool:
        """Return whether this rule covers the category."""
        if not self.categories:
            return True
        return any(str(c) == str(category) for c in self.categories)


@lru_cache(maxsize=256)
def glob_to_regexp(pattern: str) -> re.Pattern[str]:
    """Compile and cache an anchored path expression with ** support for repeated finding checks."""
    parts = ["^"]
    i = 0
    while i < len(pattern):
        if pattern.startswith("**", i):
            parts.append(".*")
            i += 2
        elif pattern[i] == "*":
            parts.append("[^/]*")
            i += 1
        elif pattern[i] == "?":
            parts.append("[^/]")
            i += 1
        else:
            parts.append(re.escape(pattern[i]))
            i += 1
    parts.append("$")
    return re.compile("".join(parts))


def apply_path_rules(findings: Sequence[Finding], rules: Sequence[PathRule]) -> list[Finding]:
    """Remove findings matching a path rule and one of its covered categories."""
    if not rules:
        return list(findings)

    out: list[Finding] = []
    for finding in findings:
        # Git and configuration use forward slashes on every platform; use posixpath for
        # matching.
        clean = posixpath.normpath(finding.file)
        if any(
            r.suppresses(finding.category) and glob_to_regexp(r.pattern).search(clean)
            for r in rules
        ):
            continue
        out.append(finding)
    return out
